cpp/include/cudf/column/column_device_view.cuh

```diff
@@ -1151,7 +1151,7 @@ struct optional_accessor {
     if (with_nulls) { CUDF_EXPECTS(_col.nullable(), "Unexpected non-nullable column."); }
   }
 
-  CUDA_DEVICE_CALLABLE
+  CUDF_DI
```

Comment:
I'm sorry, but I hate this change. This is going from something perfectly clear "CUDF_DEVICE_CALLABLE" -- a function callable from device code. to something perfectly unclear. "CUDF_DI"? What does that even mean? Is looks like a typo for "CUDF_DIE".

Reply:
In the interest of not fragmenting this discussion, even though it's a thread on an already closed PR can we continue this debate at https://github.com/rapidsai/cudf/pull/9530#discussion_r739597422? That way all of the arguments are contained in one place.

Reply:
@harrism it's one of those things where its confusing the first time you see it, but after you learn it the first time, you remember it forever. `CUDF_DI` is vastly preferable to `CUDF_HOST_DEVICE_FORCEINLINE_CALLABLE`.

Reply:
I commented here because clicking that comment link just takes me to the PR, not the comment. I'll have to go searching for it...

Reply:
I commented there, but I think it's dumb to have an unresolved discussion in a *resolved* comment on a closed PR.  So I'm repeating my comment here: 

I suggest dropping "CALLABLE" because it's easily inferred that this is a function modifier.  So I guess I'm agreeing with @bdice. I don't like "DI", "HDFI", etc.

I'll also offer that forceinline should probably be in a separate macro. After all, it's orthogonal to `__host__` and __device__`. Moreover, [there are multiple ways to spell "force inline" for different compilers](https://stackoverflow.com/a/2765211/749748), and I think separating it would make a flexible macro more manageable, should the need arise. A separate macro would also avoid making the current macros too long and since there will be a space between them it won't cause long lines.

Reply:
@harrism Unfortunately the inlining is not orthogonal because we define everything as `inline` and some functions need `__forceinline__` instead. (I had wished for the same thing.)
https://github.com/rapidsai/cudf/pull/9530#discussion_r737900785

Quoting @vyasr:
> We can't use `CUDA_HOST_DEVICE_CALLABLE` here because it embeds `inline`: https://github.com/rapidsai/cudf/blob/branch-21.12/cpp/include/cudf/types.hpp#L20. The compiler prohibits qualifying a function with both `inline` and `__forceinline__`, I assume because at some levels in the compilation they are treated as synonymous.

Reply:
> I'll also offer that forceinline should probably be in a separate macro.

`CUDF_HOST_DEVICE CUDF_FORCEINLINE` is just as bad as `CUDF_HOST_DEVICE_FORCEINLINE_CALLABLE`. 

I'd be fine with `CUDF_HD __forceinline__` though. The fact that `inline` was bundled with `HOST_DEVICE_CALLABLE` is a historical artifact and not a requirement. 

`DI/HDFI` are like `e.g.` or `i.e.`. The first time you see them you may not know what they mean, but once you've learned them the first time, you remember them in future uses (though some people have trouble remembering which is which, but that's orthogonal). It would be cumbersome/silly to spell out "_exempli gratia_" or even "for example" every place that `e.g.` is used. 

If we want a stronger reminder of what these are for, we could add leading/trailing underscores to make them look more like `__host__ __device__`:
```
_CUDF_DI_
__CUDF_DI__
```

Reply:
> CUDF_HOST_DEVICE CUDF_FORCEINLINE is just as bad as CUDF_HOST_DEVICE_FORCEINLINE_CALLABLE

![image](https://user-images.githubusercontent.com/783069/146090251-12284692-10a4-4282-b3be-e0306da94505.png)

It's not just as bad, because it can be broken across two lines if needed. 

To @bdice: if we are refactoring the macros, why not remove `inline` from the existing macros?

Reply:
> To @bdice: if we are refactoring the macros, why not remove inline from the existing macros?

I’m fine with that! I was just echoing the previous conversation before removing inline was proposed.

Reply:
> It's not just as bad, because it can be broken across two lines if needed.

It is just as bad because that's not where the line breaks are going to happen. It's going to force worse formatting of the function name/parameters/etc. 

To summarize what I said above, I think the only macro that needs to exist is `CUDF_HD`. `inline` and `__forceinline__` can and should be separate.

Reply:
In which case what's wrong with `CUDF_HOST_DEVICE`?

Reply:
I agree that `__host__ __device__` is the only case that actually needs a conditional macro since that's the only one that will be used in files that could be compiled by either the host or device compiler. I'll take @jrhemstad's 👍 on the last comment to indicate approval for `CUDF_HOST_DEVICE`.

I'm in favor of removing inlining entirely from the macros, but I'm wary of actually changing the inlining behavior of functions as part of this PR without assessing performance impacts, so I'm just going to replace existing instances of `CUDA_HOST_DEVICE_CALLABLE` with `CUDF_HOST_DEVICE inline` in this PR and we can make a separate pass to address inlining changes. Let me know if anyone is unhappy with that choice.

Reply:
I am happy with this solution! Separating `inline` out of the macro and only defining `CUDF_HOST_DEVICE` seems like a good outcome.

Reply:
Agree!